vmm/src/lib.rs

```diff
@@ -2166,95 +2273,33 @@ impl RequestHandler for Vmm {
-        let mut existing_memory_files = vec![];
-        loop {
-            let req = Request::read_from(&mut socket)?;
-            match req.command() {
-                Command::Invalid => info!("Invalid Command Received"),
-                Command::Start => {
-                    info!("Start Command Received");
-                    started = true;
-
-                    Response::ok().write_to(&mut socket)?;
-                }
-                Command::Config => {
-                    info!("Config Command Received");
+        let mut state = ReceiveMigrationState::Established;
 
-                    if !started {
-                        warn!("Migration not started yet");
-                        Response::error().write_to(&mut socket)?;
-                        continue;
-                    }
-                    memory_manager = Some(self.vm_receive_config(
-                        &req,
-                        &mut socket,
-                        HashMap::from_iter(existing_memory_files.drain(..)),
-                    )?);
-                }
-                Command::State => {
-                    info!("State Command Received");
-
-                    if !started {
-                        warn!("Migration not started yet");
-                        Response::error().write_to(&mut socket)?;
-                        continue;
-                    }
-                    if let Some(mm) = memory_manager.take() {
-                        self.vm_receive_state(&req, &mut socket, mm)?;
-                    } else {
-                        warn!("Configuration not sent yet");
-                        Response::error().write_to(&mut socket)?;
-                    }
-                }
-                Command::Memory => {
-                    info!("Memory Command Received");
+        while !state.finished() {
+            let req = Request::read_from(&mut socket)?;
+            trace!("Command {:?} received", req.command());
 
-                    if !started {
-                        warn!("Migration not started yet");
-                        Response::error().write_to(&mut socket)?;
-                        continue;
-                    }
-                    if let Some(mm) = memory_manager.as_ref() {
-                        self.vm_receive_memory(&req, &mut socket, &mut mm.lock().unwrap())?;
-                    } else {
-                        warn!("Configuration not sent yet");
-                        Response::error().write_to(&mut socket)?;
-                    }
+            let (response, new_state) = match self.vm_receive_migration_step(
+                &mut socket,
+                state,
+                &req,
+                &receive_data_migration,
+            ) {
+                Ok(next_state) => (Response::ok(), next_state),
+                Err(err) => {
+                    warn!("Migration command {:?} failed: {}", req.command(), err);
```

Comment:
Let's also make the warning message clear that the migration is now aborted (from the destination VMM side).